BrainPortal/lib/boutiques_input_copier.rb

```diff
@@ -0,0 +1,293 @@
+
+#
+# CBRAIN Project
+#
+# Copyright (C) 2008-2024
+# The Royal Institution for the Advancement of Learning
+# McGill University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+#
+# Some tools modify the cached version of the input directory.
+#
+# This can lead to some issues when the input is used multiple times
+# on the same Bourreau and the cached version is already modified.
+#
+# To solve this issue, we can create a copy of the directory in the working
+# directory and use this copy in the command line.
+#
+#     apptool [DIRECTORY_INPUT] ...
+#
+# with the following input in Boutiques:
+#
+#     {
+#       "description": "A directory that contain the input data...",
+#       "id": "file_input_id",
+#       "name": "Input",
+#       "optional": true,
+#       "type": "File",
+#       "command-line-flag": "-i",
+#       "value-key": "<-i input>"
+#     },
+#
+# and in the `cbrain:integrator_modules` section look like:
+#
+#     "BoutiquesInputCopier": {
+#       "file_input_id": {"select_box_selected": boolean, "select_box_hidden": boolean, "select_box_description": string},
+#      },
+#
+# select_box_selected:    boolean, true if the parent_input should be copied by default
+# select_box_hidden:      boolean, true if the select_box should be hidden in the form,
+# if set to true the input_file will be copied in all case.
+# select_box_description: string, description of the select_box input
+#
+# If "select_box_hidden" is set to false, a checkbox (select_box) will be added to the form
+# to allow the user to choose if the input should be copied or not.
+#
+# If the checkbox is selected, the input will be copied in the working directory replacing the symlink
+# previously created
+#
+#     apptool input_copy
+#
+# WARNING: Since this module create a copy of the input and remove
+# the original symlink, it make the task not-restartable, it will fail
+# during the setup phase of the restarted task if the input is already copied.
+#
+module BoutiquesInputCopier
+
+  # Note: to access the revision info of the module,
+  # you need to access the constant directly, the
+  # object method revision_info() won't work.
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+
+  ############################################
+  # Portal Side Modifications
+  ############################################
+
+
+  # Utility Methods for the Module
```

Comment:
General comments

* reorder the methods, the small utility methods at the bottom, the important code at the top
* get_descriptor_id() doesn't `get` anything, and the name is way too generic. Use a better name, e.g. `create_id_for_checkbox()`